From a01e292184e9389c264377fd9524e88ca97d88b7 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Wed, 6 Aug 2025 15:57:51 +0200
Subject: [PATCH 21/25] board: solidrun: lx2160cex7: rewrite fdt fixups for
 serdes ports

Serdes protocol numbers are unreliabel source of information since
they are now customizable from RCW PBI section.

Instead rely on the various functional blocks configuration registers to
check each interface status individually.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 board/solidrun/lx2160acex7/Makefile          |   1 +
 board/solidrun/lx2160acex7/eth_lx2160acex7.c | 312 -----------
 board/solidrun/lx2160acex7/lx2160a.c         |  28 +-
 board/solidrun/lx2160acex7/serdes.c          | 512 +++++++++++++++++++
 4 files changed, 533 insertions(+), 320 deletions(-)
 create mode 100644 board/solidrun/lx2160acex7/serdes.c

diff --git a/board/solidrun/lx2160acex7/Makefile b/board/solidrun/lx2160acex7/Makefile
index 4a3b039a5cf..a4bcd539cf8 100644
--- a/board/solidrun/lx2160acex7/Makefile
+++ b/board/solidrun/lx2160acex7/Makefile
@@ -8,3 +8,4 @@
 obj-y += lx2160a.o
 obj-y += ddr.o
 obj-$(CONFIG_TARGET_LX2160ACEX7) += eth_lx2160acex7.o
+obj-y += serdes.o
diff --git a/board/solidrun/lx2160acex7/eth_lx2160acex7.c b/board/solidrun/lx2160acex7/eth_lx2160acex7.c
index cb8cd425c68..bc7ae231c58 100644
--- a/board/solidrun/lx2160acex7/eth_lx2160acex7.c
+++ b/board/solidrun/lx2160acex7/eth_lx2160acex7.c
@@ -43,315 +43,3 @@ int do_mac(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
 	puts("Not implemented.\n");
 	return CMD_RET_FAILURE;
 }
-
-static void dpmac_set_phymode(void *fdt, unsigned int id, const char *mode) {
-	char path[34] = {};
-	int node;
-
-	snprintf(path, sizeof(path), "/fsl-mc@80c000000/dpmacs/dpmac@%x", id);
-	node = fdt_path_offset(fdt, path);
-	fdt_delprop(fdt, node, "phy-mode");
-	do_fixup_by_path_string(fdt, path, "phy-connection-type", mode);
-	do_fixup_by_path_string(fdt, path, "status", "okay");
-}
-
-/*
- * Fixup dpmac phy-modes by serdes protocol to fix ethernet driver probe
- */
-static void board_fix_fdt_dpmac_phymode(void *fdt, u32 srds_s1, u32 srds_s2, u32 is_lx2162) {
-	switch (srds_s1) {
-	case 0:
-	case 1:
-		break;
-	case 2:
-		/* 3, 4, 5, 6 = sgmii */
-		dpmac_set_phymode(fdt, 3, "sgmii");
-		dpmac_set_phymode(fdt, 4, "sgmii");
-		dpmac_set_phymode(fdt, 5, "sgmii");
-		dpmac_set_phymode(fdt, 6, "sgmii");
-		break;
-	case 3:
-		/* 3, 4, 5, 6 = xgmii */
-		dpmac_set_phymode(fdt, 3, "xgmii");
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "xgmii");
-		dpmac_set_phymode(fdt, 6, "xgmii");
-		break;
-	case 4:
-		/* 3, 4, 5, 6, 7, 8, 9, 10 = sgmii */
-		dpmac_set_phymode(fdt, 3, "sgmii");
-		dpmac_set_phymode(fdt, 4, "sgmii");
-		dpmac_set_phymode(fdt, 5, "sgmii");
-		dpmac_set_phymode(fdt, 6, "sgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 7, "sgmii");
-		dpmac_set_phymode(fdt, 8, "sgmii");
-		dpmac_set_phymode(fdt, 9, "sgmii");
-		dpmac_set_phymode(fdt, 10, "sgmii");
-		break;
-	case 5:
-		/* 7, 8, 9, 10 = xgmii */
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 7, "xgmii");
-		dpmac_set_phymode(fdt, 8, "xgmii");
-		dpmac_set_phymode(fdt, 9, "xgmii");
-		dpmac_set_phymode(fdt, 10, "xgmii");
-		break;
-	case 6:
-		/* 3, 4 = xgmii; 5, 6, 7, 8, 9, 10 = sgmii */
-		dpmac_set_phymode(fdt, 3, "xgmii");
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "sgmii");
-		dpmac_set_phymode(fdt, 6, "sgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 7, "sgmii");
-		dpmac_set_phymode(fdt, 8, "sgmii");
-		dpmac_set_phymode(fdt, 9, "sgmii");
-		dpmac_set_phymode(fdt, 10, "sgmii");
-		break;
-	case 7:
-		/* 3, 4, 5, 6 = xgmii; 7, 8, 9, 10 = sgmii */
-		dpmac_set_phymode(fdt, 3, "xgmii");
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "xgmii");
-		dpmac_set_phymode(fdt, 6, "xgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 7, "sgmii");
-		dpmac_set_phymode(fdt, 8, "sgmii");
-		dpmac_set_phymode(fdt, 9, "sgmii");
-		dpmac_set_phymode(fdt, 10, "sgmii");
-		break;
-	case 8:
-		/* 3, 4, 5, 6, 7, 8, 9, 10 = xgmii */
-		dpmac_set_phymode(fdt, 3, "xgmii");
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "xgmii");
-		dpmac_set_phymode(fdt, 6, "xgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 7, "xgmii");
-		dpmac_set_phymode(fdt, 8, "xgmii");
-		dpmac_set_phymode(fdt, 9, "xgmii");
-		dpmac_set_phymode(fdt, 10, "xgmii");
-		break;
-	case 9:
-		/* 4, 5, 6, 8, 9, 10 = sgmii */
-		dpmac_set_phymode(fdt, 4, "sgmii");
-		dpmac_set_phymode(fdt, 5, "sgmii");
-		dpmac_set_phymode(fdt, 6, "sgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 8, "sgmii");
-		dpmac_set_phymode(fdt, 9, "sgmii");
-		dpmac_set_phymode(fdt, 10, "sgmii");
-		break;
-	case 10:
-		/* 4, 5, 6, 8, 9, 10 = xgmii */
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "xgmii");
-		dpmac_set_phymode(fdt, 6, "xgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 8, "xgmii");
-		dpmac_set_phymode(fdt, 9, "xgmii");
-		dpmac_set_phymode(fdt, 10, "xgmii");
-		break;
-	case 11:
-		/* 5, 6, 9, 10 = sgmii */
-		dpmac_set_phymode(fdt, 5, "sgmii");
-		dpmac_set_phymode(fdt, 6, "sgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 9, "sgmii");
-		dpmac_set_phymode(fdt, 10, "sgmii");
-		break;
-	case 12:
-		/* 9, 10 = sgmii */
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 9, "sgmii");
-		dpmac_set_phymode(fdt, 10, "sgmii");
-		break;
-	case 13:
-		/* 1, 2 = caui4 */
-		dpmac_set_phymode(fdt, 1, "caui4");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 2, "caui4");
-		break;
-	case 14:
-		/* 1 = caui4 */
-		dpmac_set_phymode(fdt, 1, "caui4");
-		break;
-	case 15:
-		/* 1, 2 = caui2 */
-		dpmac_set_phymode(fdt, 1, "caui2");
-		dpmac_set_phymode(fdt, 2, "caui2");
-		break;
-	case 16:
-		/* 1 = caui2; 5, 6 = 25g-aui */
-		dpmac_set_phymode(fdt, 1, "caui2");
-		dpmac_set_phymode(fdt, 5, "25g-aui");
-		dpmac_set_phymode(fdt, 6, "25g-aui");
-		break;
-	case 17:
-		/* 3, 4, 5, 6 = 25g-aui */
-		dpmac_set_phymode(fdt, 3, "25g-aui");
-		dpmac_set_phymode(fdt, 4, "25g-aui");
-		dpmac_set_phymode(fdt, 5, "25g-aui");
-		dpmac_set_phymode(fdt, 6, "25g-aui");
-		break;
-	case 18:
-		/* 3, 4, 7, 8, 9, 10 = xgmii; 5, 6 = 25g-aui */
-		dpmac_set_phymode(fdt, 3, "xgmii");
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "25g-aui");
-		dpmac_set_phymode(fdt, 6, "25g-aui");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 7, "xgmii");
-		dpmac_set_phymode(fdt, 8, "xgmii");
-		dpmac_set_phymode(fdt, 9, "xgmii");
-		dpmac_set_phymode(fdt, 10, "xgmii");
-		break;
-	case 19:
-		/* 2 = xlaui4; 3, 4 = xgmii; 5, 6 = 25g-aui */
-		dpmac_set_phymode(fdt, 3, "xgmii");
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "25g-aui");
-		dpmac_set_phymode(fdt, 6, "25g-aui");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 2, "xlaui4");
-		break;
-	case 20:
-		/* 1, 2 = xlaui4 */
-		dpmac_set_phymode(fdt, 1, "xlaui4");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 2, "xlaui4");
-		break;
-	case 21:
-		/* 3, 4, 5, 6, 9, 10 = 25g-aui */
-		dpmac_set_phymode(fdt, 3, "25g-aui");
-		dpmac_set_phymode(fdt, 4, "25g-aui");
-		dpmac_set_phymode(fdt, 5, "25g-aui");
-		dpmac_set_phymode(fdt, 6, "25g-aui");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 9, "25g-aui");
-		dpmac_set_phymode(fdt, 10, "25g-aui");
-		break;
-	case 22:
-		/* 3, 4, 5, 6, 9, 10 = xgmii */
-		dpmac_set_phymode(fdt, 3, "xgmii");
-		dpmac_set_phymode(fdt, 4, "xgmii");
-		dpmac_set_phymode(fdt, 5, "xgmii");
-		dpmac_set_phymode(fdt, 6, "xgmii");
-		if (is_lx2162)
-			break;
-		dpmac_set_phymode(fdt, 9, "xgmii");
-		dpmac_set_phymode(fdt, 10, "xgmii");
-		break;
-	}
-
-	switch (srds_s2) {
-	case 0:
-	case 1:
-	case 2:
-	case 3:
-	case 4:
-	case 5:
-		break;
-	case 6:
-		/* 13, 14 = xgmii; 15, 16 = sgmii */
-		dpmac_set_phymode(fdt, 13, "xgmii");
-		dpmac_set_phymode(fdt, 14, "xgmii");
-		dpmac_set_phymode(fdt, 15, "sgmii");
-		dpmac_set_phymode(fdt, 16, "sgmii");
-		break;
-	case 7:
-		/* 12, 16, 17, 18 = sgmii; 13, 14 = xgmii */
-		dpmac_set_phymode(fdt, 12, "sgmii");
-		dpmac_set_phymode(fdt, 13, "xgmii");
-		dpmac_set_phymode(fdt, 14, "xgmii");
-		dpmac_set_phymode(fdt, 16, "sgmii");
-		dpmac_set_phymode(fdt, 17, "sgmii");
-		dpmac_set_phymode(fdt, 18, "sgmii");
-		break;
-	case 8:
-		/* 13, 14 = xgmii */
-		dpmac_set_phymode(fdt, 13, "xgmii");
-		dpmac_set_phymode(fdt, 14, "xgmii");
-		break;
-	case 9:
-		/* 11, 12, 13, 14, 15, 16, 17, 18 = sgmii */
-		dpmac_set_phymode(fdt, 11, "sgmii");
-		dpmac_set_phymode(fdt, 12, "sgmii");
-		dpmac_set_phymode(fdt, 13, "sgmii");
-		dpmac_set_phymode(fdt, 14, "sgmii");
-		dpmac_set_phymode(fdt, 15, "sgmii");
-		dpmac_set_phymode(fdt, 16, "sgmii");
-		dpmac_set_phymode(fdt, 17, "sgmii");
-		dpmac_set_phymode(fdt, 18, "sgmii");
-		break;
-	case 10:
-		/* 11, 12, 17, 18 = sgmii */
-		dpmac_set_phymode(fdt, 11, "sgmii");
-		dpmac_set_phymode(fdt, 12, "sgmii");
-		dpmac_set_phymode(fdt, 17, "sgmii");
-		dpmac_set_phymode(fdt, 18, "sgmii");
-		break;
-	case 11:
-		/* 12, 13, 14, 16, 17, 18 = sgmii */
-		dpmac_set_phymode(fdt, 12, "sgmii");
-		dpmac_set_phymode(fdt, 13, "sgmii");
-		dpmac_set_phymode(fdt, 14, "sgmii");
-		dpmac_set_phymode(fdt, 16, "sgmii");
-		dpmac_set_phymode(fdt, 17, "sgmii");
-		dpmac_set_phymode(fdt, 18, "sgmii");
-		break;
-	case 12:
-		/* 11, 12, 17, 18 = sgmii */
-		dpmac_set_phymode(fdt, 11, "sgmii");
-		dpmac_set_phymode(fdt, 12, "sgmii");
-		dpmac_set_phymode(fdt, 17, "sgmii");
-		dpmac_set_phymode(fdt, 18, "sgmii");
-		break;
-	case 13:
-		/* 13, 14 = sgmii */
-		dpmac_set_phymode(fdt, 13, "sgmii");
-		dpmac_set_phymode(fdt, 14, "sgmii");
-		break;
-	case 14:
-		/* 13, 14, 17, 18 = sgmii */
-		dpmac_set_phymode(fdt, 13, "sgmii");
-		dpmac_set_phymode(fdt, 14, "sgmii");
-		dpmac_set_phymode(fdt, 17, "sgmii");
-		dpmac_set_phymode(fdt, 18, "sgmii");
-		break;
-	}
-}
-
-void board_fix_fdt_eth(void *fdt) {
-	struct ccsr_gur __iomem *gur = (void *)(CFG_SYS_FSL_GUTS_ADDR);
-	u32 is_lx2162 = get_svr() & 0x800;
-	u32 srds_s1, srds_s2;
-
-	srds_s1 = in_le32(&gur->rcwsr[FSL_CHASSIS3_SRDS1_REGSR - 1]) &
-			  FSL_CHASSIS3_RCWSR28_SRDS1_PRTCL_MASK;
-	srds_s1 >>= FSL_CHASSIS3_RCWSR28_SRDS1_PRTCL_SHIFT;
-
-	srds_s2 = in_le32(&gur->rcwsr[FSL_CHASSIS3_SRDS2_REGSR - 1]) &
-			  FSL_CHASSIS3_RCWSR28_SRDS2_PRTCL_MASK;
-	srds_s2 >>= FSL_CHASSIS3_RCWSR28_SRDS2_PRTCL_SHIFT;
-
-	/* allocate space in case properties must be added */
-	fdt_increase_size(fdt, 256);
-
-	board_fix_fdt_dpmac_phymode(fdt, srds_s1, srds_s2, is_lx2162);
-}
diff --git a/board/solidrun/lx2160acex7/lx2160a.c b/board/solidrun/lx2160acex7/lx2160a.c
index 1f742b0b0c6..632fbff52dc 100644
--- a/board/solidrun/lx2160acex7/lx2160a.c
+++ b/board/solidrun/lx2160acex7/lx2160a.c
@@ -17,7 +17,7 @@
 #include <fsl-mc/fsl_mc.h>
 #include <fsl_ddr.h>
 #include <init.h>
-#include <malloc.h>>
+#include <malloc.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -28,11 +28,10 @@ int board_early_init_f(void)
 }
 
 #ifdef CONFIG_OF_BOARD_FIXUP
-void board_fix_fdt_eth(void *fdt);
+void board_fix_fdt_eth(void *fdt, u32 srds_s1, u32 srds_s2, u32 is_lx2162);
+void board_fix_fdt_serdes_ports(void *fdt);
 
-/* fdt fixup for u-boot itself */
-int board_fix_fdt(void *fdt)
-{
+static void board_fix_fdt_pci_silicon(void *fdt) {
 	char *reg_names, *reg_name;
 	int names_len, old_name_len, new_name_len, remaining_names_len;
 	struct str_map {
@@ -44,11 +43,9 @@ int board_fix_fdt(void *fdt)
 	};
 	int off = -1, i = 0;
 
-	board_fix_fdt_eth(fdt);
-
 	/* skip pci fixup on silicon version 1 */
 	if (IS_SVR_REV(get_svr(), 1, 0))
-		return 0;
+		return;
 
 	/* fixup pci controller for silicon version 2 */
 	off = fdt_node_offset_by_compatible(fdt, -1, "fsl,lx2160a-pcie");
@@ -97,6 +94,17 @@ int board_fix_fdt(void *fdt)
 		off = fdt_node_offset_by_compatible(fdt, off,
 						    "fsl,lx2160a-pcie");
 	}
+}
+
+/* fdt fixup for u-boot itself */
+int board_fix_fdt(void *fdt)
+{
+	/* allocate space in case properties must be added */
+	fdt_increase_size(fdt, 512);
+
+	/* fix fdt */
+	board_fix_fdt_serdes_ports(fdt);
+	board_fix_fdt_pci_silicon(fdt);
 
 	return 0;
 }
@@ -135,6 +143,10 @@ unsigned long get_board_ddr_clk(void)
 
 int board_init(void)
 {
+#if !defined(CONFIG_SYS_EARLY_PCI_INIT)
+	pci_init();
+#endif
+
 	return 0;
 }
 
diff --git a/board/solidrun/lx2160acex7/serdes.c b/board/solidrun/lx2160acex7/serdes.c
new file mode 100644
index 00000000000..c6b6d9bbfdc
--- /dev/null
+++ b/board/solidrun/lx2160acex7/serdes.c
@@ -0,0 +1,512 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2025 Josua Mayer <josua@solid-run.com>
+ *
+ */
+
+#include <fdt_support.h>
+
+/* SerDes base address */
+#define LYNX_28G_SDn_BASE(block)		((void *)0x01EA0000 + (block) * 0x10000)
+
+/* Protocol Configuration Register 0 */
+#define LYNX_28G_PCC0				0x1080
+#define LYNX_28G_PCC0_PEXA_CFG			GENMASK(30, 28)
+#define LYNX_28G_PCC0_PEXB_CFG			GENMASK(26, 24)
+
+/* Protocol Configuration Register 2 */
+#define LYNX_28G_PCC2				0x1088
+#define LYNX_28G_PCC2_SATAA_CFG			GENMASK(30, 28)
+#define LYNX_28G_PCC2_SATAB_CFG			GENMASK(26, 24)
+#define LYNX_28G_PCC2_SATAC_CFG			GENMASK(22, 20)
+#define LYNX_28G_PCC2_SATAD_CFG			GENMASK(18, 16)
+
+/* Protocol Configuration Register 8 */
+#define LYNX_28G_PCC8				0x10A0
+#define LYNX_28G_PCC8_SGMIIA_CFG		GENMASK(30, 28)
+#define LYNX_28G_PCC8_SGMIIB_CFG		GENMASK(26, 24)
+#define LYNX_28G_PCC8_SGMIIC_CFG		GENMASK(22, 20)
+#define LYNX_28G_PCC8_SGMIID_CFG		GENMASK(18, 16)
+#define LYNX_28G_PCC8_SGMIIE_CFG		GENMASK(14, 12)
+#define LYNX_28G_PCC8_SGMIIF_CFG		GENMASK(10, 8)
+#define LYNX_28G_PCC8_SGMIIG_CFG		GENMASK(6, 4)
+#define LYNX_28G_PCC8_SGMIIH_CFG		GENMASK(2, 0)
+
+/* Protocol Configuration Register C */
+#define LYNX_28G_PCCC				0x10B0
+#define LYNX_28G_PCCC_SXGMIIA_CFG		GENMASK(30, 28)
+#define LYNX_28G_PCCC_SXGMIIB_CFG		GENMASK(26, 24)
+#define LYNX_28G_PCCC_SXGMIIC_CFG		GENMASK(22, 20)
+#define LYNX_28G_PCCC_SXGMIID_CFG		GENMASK(18, 16)
+#define LYNX_28G_PCCC_SXGMIIE_CFG		GENMASK(14, 12)
+#define LYNX_28G_PCCC_SXGMIIF_CFG		GENMASK(10, 8)
+#define LYNX_28G_PCCC_SXGMIIG_CFG		GENMASK(6, 4)
+#define LYNX_28G_PCCC_SXGMIIH_CFG		GENMASK(2, 0)
+
+/* Protocol Configuration Register D */
+#define LYNX_28G_PCCD				0x10B4
+#define LYNX_28G_PCCD_E25GA_CFG			GENMASK(30, 28)
+#define LYNX_28G_PCCD_E25GB_CFG			GENMASK(26, 24)
+#define LYNX_28G_PCCD_E25GC_CFG			GENMASK(22, 20)
+#define LYNX_28G_PCCD_E25GD_CFG			GENMASK(18, 16)
+#define LYNX_28G_PCCD_E25GE_CFG			GENMASK(14, 12)
+#define LYNX_28G_PCCD_E25GF_CFG			GENMASK(10, 8)
+#define LYNX_28G_PCCD_E25GG_CFG			GENMASK(6, 4)
+#define LYNX_28G_PCCD_E25GH_CFG			GENMASK(2, 0)
+
+/* Protocol Configuration Register E */
+#define LYNX_28G_PCCE				0x10B8
+#define LYNX_28G_PCCE_E40GA_CFG			GENMASK(30, 28)
+#define LYNX_28G_PCCE_E40GB_CFG			GENMASK(26, 24)
+#define LYNX_28G_PCCE_E50GA_CFG			GENMASK(22, 20)
+#define LYNX_28G_PCCE_E50GB_CFG			GENMASK(18, 16)
+#define LYNX_28G_PCCE_E100GA_CFG		GENMASK(14, 12)
+#define LYNX_28G_PCCE_E100GB_CFG		GENMASK(10, 8)
+
+/* Lane a General Control Register */
+#define LYNX_28G_LNaGCR0(lane)			(0x800 + (lane) * 0x100 + 0x0)
+#define LYNX_28G_LNaGCR0_PORT_RST_LEFT		BIT(17)
+#define LYNX_28G_LNaGCR0_PORT_LN0_B		BIT(16)
+#define LYNX_28G_LNaGCR0_PROTO_SEL_MSK		GENMASK(7, 3)
+#define LYNX_28G_LNaGCR0_PROTO_SEL_PCI		0x0
+#define LYNX_28G_LNaGCR0_PROTO_SEL_SGMII	0x8
+#define LYNX_28G_LNaGCR0_PROTO_SEL_SATA		0x10
+#define LYNX_28G_LNaGCR0_PROTO_SEL_XFI		0x50
+#define LYNX_28G_LNaGCR0_PROTO_SEL_25G		0xD0
+
+/* Lane a Tx Reset Control Register */
+#define LYNX_28G_LNaTRSTCTL(lane)		(0x800 + (lane) * 0x100 + 0x20)
+#define LYNX_28G_LNaTRSTCTL_DIS			BIT(24)
+
+/* Lane a Rx Reset Control Register */
+#define LYNX_28G_LNaRRSTCTL(lane)		(0x800 + (lane) * 0x100 + 0x40)
+#define LYNX_28G_LNaRRSTCTL_DIS			BIT(24)
+
+/* Reset Control Word 27 (RO) */
+#define RCWSR27_R				((void *)0x01e00168)
+#define RCWSR27_EC1_PMUX_MASK			0x00000003
+#define RCWSR27_EC1_PMUX_WRIOP_MAC_17_RGMII	0x00000000
+#define RCWSR27_EC2_PMUX_MASK			0x0000000C
+#define RCWSR27_EC2_PMUX_WRIOP_MAC_18_RGMII	0x00000000
+
+enum {
+	DPMAC1 = 0,
+	DPMAC2,
+	DPMAC3,
+	DPMAC4,
+	DPMAC5,
+	DPMAC6,
+	DPMAC7,
+	DPMAC8,
+	DPMAC9,
+	DPMAC10,
+	DPMAC11,
+	DPMAC12,
+	DPMAC13,
+	DPMAC14,
+	DPMAC15,
+	DPMAC16,
+	DPMAC17,
+	DPMAC18,
+	DPMAC_MAX
+};
+
+#ifdef CONFIG_OF_BOARD_FIXUP
+
+/* fix mac nodes based on serdes protocol */
+static void board_fix_fdt_macs(void *fdt) {
+	struct mac_node {
+		const char *const path;
+		const char *status;
+		const char *mode;
+	} macs[DPMAC_MAX] = {
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@1", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@2", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@3", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@4", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@5", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@6", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@7", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@8", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@9", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@a", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@b", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@c", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@d", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@e", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@f", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@10", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@11", "disabled" },
+		{ "/fsl-mc@80c000000/dpmacs/dpmac@12", "disabled" },
+	};
+
+	struct {
+		const u32 __iomem *pcr;
+		const u32 pcr_ena_mask;
+		const char *const mode;
+		const unsigned int mac;
+	} ports[] = {
+		{
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIA_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC3,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIB_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC4,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIC_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC5,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIID_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC6,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIE_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC7,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIF_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC8,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIG_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC9,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIH_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC10,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIA_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC3,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIB_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC4,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIC_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC5,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIID_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC6,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIE_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC7,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIF_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC8,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIG_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC9,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIH_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC10,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GA_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC3,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GB_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC4,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GC_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC5,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GD_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC6,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GE_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC7,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GF_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC8,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GG_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC9,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GH_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC10,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCE,
+			.pcr_ena_mask = LYNX_28G_PCCE_E40GA_CFG,
+			.mode = "xlaui4",
+			.mac = DPMAC1,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCE,
+			.pcr_ena_mask = LYNX_28G_PCCE_E40GB_CFG,
+			.mode = "xlaui4",
+			.mac = DPMAC2,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCE,
+			.pcr_ena_mask = LYNX_28G_PCCE_E50GA_CFG,
+			.mode = "caui2",
+			.mac = DPMAC1,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCE,
+			.pcr_ena_mask = LYNX_28G_PCCE_E50GB_CFG,
+			.mode = "caui2",
+			.mac = DPMAC2,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCE,
+			.pcr_ena_mask = LYNX_28G_PCCE_E100GA_CFG,
+			.mode = "caui4",
+			.mac = DPMAC1,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCCE,
+			.pcr_ena_mask = LYNX_28G_PCCE_E100GB_CFG,
+			.mode = "caui4",
+			.mac = DPMAC2,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIA_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC18,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIB_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC17,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIC_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC16,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIID_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC15,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIE_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC14,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIF_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC13,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIG_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC12,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC8,
+			.pcr_ena_mask = LYNX_28G_PCC8_SGMIIH_CFG,
+			.mode = "sgmii",
+			.mac = DPMAC11,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIA_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC18,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIB_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC17,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIC_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC16,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIID_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC15,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIE_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC14,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIF_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC13,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIG_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC12,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCC,
+			.pcr_ena_mask = LYNX_28G_PCCC_SXGMIIH_CFG,
+			.mode = "xgmii",
+			.mac = DPMAC11,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GA_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC18,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GB_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC17,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GC_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC16,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GD_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC15,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GE_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC14,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GF_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC13,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GG_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC12,
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCCD,
+			.pcr_ena_mask = LYNX_28G_PCCD_E25GH_CFG,
+			.mode = "25g-aui",
+			.mac = DPMAC11,
+		}
+	};
+
+	for (int i = 0; i < ARRAY_SIZE(ports); i++) {
+		if (*ports[i].pcr & ports[i].pcr_ena_mask) {
+			macs[ports[i].mac].status = "okay";
+			macs[ports[i].mac].mode = ports[i].mode;
+		}
+	}
+
+	/* workaround for dpmac17.18 rgmii ports */
+	{
+		const u32 __iomem *const rcwsr27 = RCWSR27_R;
+		if ((*rcwsr27 & RCWSR27_EC1_PMUX_MASK) == RCWSR27_EC1_PMUX_WRIOP_MAC_17_RGMII) {
+			macs[DPMAC17].status = "okay";
+			macs[DPMAC17].mode = "rgmii-id";
+		}
+		if ((*rcwsr27 & RCWSR27_EC2_PMUX_MASK) == RCWSR27_EC2_PMUX_WRIOP_MAC_18_RGMII) {
+			macs[DPMAC18].status = "okay";
+			macs[DPMAC18].mode = "rgmii-id";
+		}
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(macs); i++) {
+		fdt_delprop(fdt, fdt_path_offset(fdt, macs[i].path), "phy-mode");
+		do_fixup_by_path_string(fdt, macs[i].path, "status", macs[i].status);
+		do_fixup_by_path_string(fdt, macs[i].path, "phy-connection-type", macs[i].mode);
+	}
+}
+
+static void board_fix_fdt_pci_sata(void *fdt) {
+	struct {
+		const u32 __iomem *pcr;
+		const u32 pcr_ena_mask;
+		const char *const path;
+	} ports[] = {
+		{
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC0,
+			.pcr_ena_mask = LYNX_28G_PCC0_PEXA_CFG,
+			.path = "/pcie@3400000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(0) + LYNX_28G_PCC0,
+			.pcr_ena_mask = LYNX_28G_PCC0_PEXB_CFG,
+			.path = "/pcie@3500000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC0,
+			.pcr_ena_mask = LYNX_28G_PCC0_PEXA_CFG,
+			.path = "/pcie@3600000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC0,
+			.pcr_ena_mask = LYNX_28G_PCC0_PEXB_CFG,
+			.path = "/pcie@3700000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(2) + LYNX_28G_PCC0,
+			.pcr_ena_mask = LYNX_28G_PCC0_PEXA_CFG,
+			.path = "/pcie@3800000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(2) + LYNX_28G_PCC0,
+			.pcr_ena_mask = LYNX_28G_PCC0_PEXB_CFG,
+			.path = "/pcie@3900000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC2,
+			.pcr_ena_mask = LYNX_28G_PCC2_SATAA_CFG,
+			.path = "/sata@3200000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC2,
+			.pcr_ena_mask = LYNX_28G_PCC2_SATAB_CFG,
+			.path = "/sata@3210000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC2,
+			.pcr_ena_mask = LYNX_28G_PCC2_SATAC_CFG,
+			.path = "/sata@3220000",
+		}, {
+			.pcr = LYNX_28G_SDn_BASE(1) + LYNX_28G_PCC2,
+			.pcr_ena_mask = LYNX_28G_PCC2_SATAD_CFG,
+			.path = "/sata@3230000",
+		},
+	};
+
+	for (int i = 0; i < ARRAY_SIZE(ports); i++) {
+		const char *status = "disabled";
+		if (*ports[i].pcr & ports[i].pcr_ena_mask)
+			status = "okay";
+
+		do_fixup_by_path_string(fdt, ports[i].path, "status", status);
+	}
+}
+
+void board_fix_fdt_serdes_ports(void *fdt) {
+	board_fix_fdt_pci_sata(fdt);
+	board_fix_fdt_macs(fdt);
+}
+
+#endif /* CONFIG_OF_BOARD_FIXUP */
-- 
2.43.0

